<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>shinyDSP vignette • shinyDSP</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="shinyDSP vignette">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">shinyDSP</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette.html">shinyDSP vignette</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>shinyDSP vignette</h1>
                        <h4 data-toc-skip class="author">Seung J.
Kim</h4>
            <address class="author_afil">
      Interstitial Lung Disease Lab, London Health Sciences
Center<br><a class="author_email" href="mailto:#"></a><a href="mailto:seung.kim@lhsc.on.ca" class="email">seung.kim@lhsc.on.ca</a>
      </address>
                              <h4 data-toc-skip class="author">Angwara
(Book) Nilanont</h4>
            <address class="author_afil">
      Interstitial Lung Disease Lab, London Health Sciences
Center<br><h4 data-toc-skip class="author">Marco
Mura</h4>
            <address class="author_afil">
      Interstitial Lung Disease Lab, London Health Sciences
CenterDivision of Respirology, Department of Medicine, Western
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:marco.mura@lhsc.on.ca" class="email">marco.mura@lhsc.on.ca</a>
      </address>
                  
            <h4 data-toc-skip class="date">Oct 2024</h4>
      

      <div class="d-none name"><code>vignette.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>shinyDSP: Analyzing and Visualizing Nanostring GeoMx DSP Data</p>
<p>shinyDSP is an intuitive Shiny application designed for the
comprehensive analysis and visualization of <a href="https://nanostring.com/products/geomx-digital-spatial-profiler/geomx-dsp-overview/" class="external-link">Nanostring
GeoMx DSP</a> data. Users can upload either demo or custom datasets,
consisting of count and sample annotation tables. The app prompts users
to select variables of interest, potential batch effects, and
confounding factors, allowing for customized exploration.</p>
<p>With shinyDSP, users can create x-y scatter plots of any combination
of categorical variables and apply user-defined cutoffs to filter
samples. The app utilizes the R package, standR <span class="citation">(<a href="#ref-liu_standr_2023">Liu et al.
2023</a>)</span>, to perform normalization using methods such as CPM,
upper quartile (Q3), or RUV4 (Remove Unwanted Variation). Users can
visualize PCA plots generated for each normalization method, color-coded
by chosen variables or batch.</p>
<p>After selecting a normalization scheme, users can identify
differentially expressed genes between specified biological groups using
limma-voom <span class="citation">(<a href="#ref-ritchie_limma_2015">Ritchie et al. 2015</a>)</span>. The app
provides “raw” output numbers in tables, generates volcano plots for all
pairwise comparisons, and displays heatmaps of the top differentially
expressed genes.</p>
<p>shinyDSP aims to provide a robust, start-to-finish analysis of GeoMx
data, producing publication-ready outputs that are easily customizable
to meet individual aesthetic preferences.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"shinyDSP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To install the development version from Github:</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"kimsjune/shinyDSP"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kimsjune.github.com/ShinyDSP" class="external-link">shinyDSP</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kimsjune.github.com/ShinyDSP" class="external-link">shinyDSP</a></span><span class="op">)</span></span>
<span><span class="va">app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shinyDSP.html">shinyDSP</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># This will open a new browser tab/window.</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">app</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="user-interface">User interface<a class="anchor" aria-label="anchor" href="#user-interface"></a>
</h2>
<p><img src="UI_annotated.png" style="display: block; margin: 1em auto;;width:100.0%"></p>
<p>There are four main UI components:<br>
1. The <code>nav bar</code> where <code>nav panels</code> will appear.
<code>Setup</code> is a <code>nav panel</code>.<br>
2. The main <code>side bar</code>. This is where you can set global
parameters.<br>
3. The main display within each <code>nav panel</code>. This is where
outputs will appear.<br>
4. A <code>side bar</code> within the <code>setup</code>
<code>nav panel</code>. This is where customization options will
appear.</p>
</div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>shinyDSP requires count and annotation tables (in .csv or .txt
format) as input. These tables are generated with DSP Data Analysis
(DSPDA) software. To see how these files should be formatted, click on
“Use <a href="https://nanostring.com/products/geomx-digital-spatial-profiler/spatial-organ-atlas/human-kidney/" class="external-link">demo
data</a>”, then “Load data”. The top 10 rows of each table will be shown
in the main display.</p>
<p>The main <code>side bar</code> will also be updated with various
options.</p>
<p><img src="main_sidebar_annotated.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<p>1. You can pick one or more variables of interest. It’s common to
combine two variables into one grouped variable. For example, “genotype”
and “treatment” can be combined into “genotype_treatment”. A new column
in your annotation table is automatically created.<br>
2. A batch variable. “SlideName” is selected by default, but it could be
any categorical variable provided in the annotation table such as
“sample preparation batch”, etc.<br>
3. Any confounding variable(s) such as age or sex of your samples that
you want to include in the design matrix for differential gene
expression analysis. None selected by default.</p>
<p>For this demo, “disease_status” and “region” are selected as the
variables of interest. All four groups of interest are selected
(DKD_glomerulus, DKD_tubule, normal_tubule, normal_glomerulus). After
selecting “Variable(s) of interest”, two new <code>nav panels</code>
appear: “QC” and “PCA”.</p>
</div>
<div class="section level2">
<h2 id="qc">QC<a class="anchor" aria-label="anchor" href="#qc"></a>
</h2>
<p>Click on the “QC” <code>nav panel</code> to create scatter plots and
(optionally) filter samples not meeting cutoffs. <img src="qc_sidebar_annotated.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<p>1. Pick <strong>two</strong> or more quantitative variables to plot
and (optionally) filter.<br>
2. If you select more than two variables, increase this number to show
all possible x-y plots.<br>
3. Pick a variable for the colour legend. “SlideName” is the
default.<br>
4. Pick one of the five colour palettes. “glasbey” is the default.<br>
5. Click to open. You have the option of providing minimum threshold
value(s) for each variable from (1).</p>
<p>Lastly, click on “Show QC plots” to show/update the plots.</p>
<p><img src="qc_main_annotated.png" style="width:100.0%"> In the
example above, “SequencingSaturation” and “DeduplicatedReads” were
selected. Then, I removed any samples with “SequencingSaturation” below
85. No filtering was applied based on “DeduplicatedReads”.</p>
<p>The scatter plot can be saved as .png, .tiff, .svg or .pdf by
clicking the download buttons below each plot.</p>
<p>Now we move on to “PCA” (click on the <code>nav panel</code>).</p>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p><img src="pca_sidebar.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<p>Click on “Run” to generate PCA plots. For each group of interest and
batch variable that you selected in the main <code>side bar</code>, you
can pick its shape and colour. For example, “DKD_glomerulus” will appear
as black circles. You can pick between five different shapes and any
colour in <a href="">grDevices::colours()</a>.</p>
<p>Two sets of three PCA plots are generated in the main display area.
Three normalization schemes are shown: CPM, Q3 (upperquartile), and RUV4
(Remove Unwanted Variation). Across the top and buttom row, the plots
are colour-coded by “Variable(s) of interest”, and the “batch variable”,
respectively. Click on “Download” in the <code>side bar</code> to find
all the download options.</p>
<p>Select the smallest value of “k value for RUV4 norm.” that removes
any batch effect. Click on “Run” to show updated plots.</p>
</div>
<div class="section level2">
<h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p><img src="main_sidebar_norm.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<p>Now you can choose the normalization scheme to use for differential
gene expression testing and the log fold-change cutoff for <a href="https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/toptable" class="external-link">limma::topTable</a>.
Selecting a normalization scheme will reveal three new
<code>nav panels</code> in the <code>nav bar</code>: “Table”, “Volcano”,
and “Heatmap”.</p>
<p><img src="updated_nav_bar.png" style="display: block; margin: 1em auto;;width:100.0%"></p>
<blockquote>
<p>If you choose “RUV4”, you need to open the “PCA”
<code>nav panel</code> to load the k value.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="table">Table<a class="anchor" aria-label="anchor" href="#table"></a>
</h2>
<p>Clicking on the “Table” <code>nav panel</code> will automatically
start performing differential gene expression testing between all
selected “Groups of interest”. If more than two groups were selected,
all possible pairwise comparisons <em>and</em> an ANOVA-like test
between all groups are executed. This step can take about 3 minutes.
Results are separated into <code>tabs</code> highlighted in blue. Click
on the “Download table” button below each table to download it (in
.csv).</p>
<p><img src="table.png" style="display: block; margin: 1em auto;;width:100.0%"></p>
</div>
<div class="section level2">
<h2 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h2>
<p>Click on the “Volcano” <code>nav panel</code> and “Show/update”.</p>
<p><img src="volcano_main.png" style="display: block; margin: 1em auto;;width:100.0%"></p>
<p>Like the tables, Volcano plots are shown in individual
<code>tabs</code> highlighted in blue.</p>
<p><img src="volcano_sidebar_annotated.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<p>There are several customization options for tweaking the look and
feel of these plots.<br>
1. Higher number increases the number of gene names shown by allowing
them to overlap each other.<br>
2. Higher number makes the labels larger.<br>
3-4. These options are used to colour those genes not meeting cutoffs to
have “Not DE colour (7)”.<br>
5. Those genes with logFC &gt;= “logFC cutoff” are given this colour.
Must be a grDevices::color().<br>
6. Those genes with logFC &lt;= -“logFC cutoff” are given this colour.
Must be a grDevices::color()“.<br>
7. Must be a grDevices::color().<br>
8. Click to enable custom x and y ranges.</p>
<p>These settings are applied to all Volcano plots. Below each Volcano
plot, you have the option to save it as four different file types.</p>
</div>
<div class="section level2">
<h2 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h2>
<p>For each “Table”, a corresponding heatmap is generated. The heatmaps
are also organized into individual <code>tabs</code>. By default, the
top 50 genes (sorted by adjusted P value) are shown as rows, clustered
based on Euclidean distances. There are a few customization options:</p>
<p><img src="heatmap_sidebar_annotated.png" style="display: block; margin: 1em auto;;width:50.0%"></p>
<ol style="list-style-type: decimal">
<li>Any N top genes can be plotted.<br>
</li>
<li>The <a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html" class="external-link">viridis</a>
colour map is available.<br>
</li>
<li>A custom range of Z score. Does not have to be balanced.<br>
</li>
<li>This adjusts the overall size of the **downloaded* heatmap. Each
“square” will become smaller/bigger.<br>
</li>
<li>Font size for row/gene labels.</li>
</ol>
<p>Below each heatmap, you have the option to save it as four different
file types.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-10-01 r87205 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 22631)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=English_Canada.utf8  LC_CTYPE=English_Canada.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=English_Canada.utf8 LC_NUMERIC=C                   </span></span>
<span><span class="co">#&gt; [5] LC_TIME=English_Canada.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Toronto</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] BiocStyle_2.33.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] digest_0.6.37       desc_1.4.3          R6_2.5.1           </span></span>
<span><span class="co">#&gt;  [4] bookdown_0.41       fastmap_1.2.0       xfun_0.48          </span></span>
<span><span class="co">#&gt;  [7] cachem_1.1.0        knitr_1.48          htmltools_0.5.8.1  </span></span>
<span><span class="co">#&gt; [10] rmarkdown_2.28      lifecycle_1.0.4     cli_3.6.3          </span></span>
<span><span class="co">#&gt; [13] sass_0.4.9          pkgdown_2.1.1       textshaping_0.4.0  </span></span>
<span><span class="co">#&gt; [16] jquerylib_0.1.4     systemfonts_1.1.0   compiler_4.5.0     </span></span>
<span><span class="co">#&gt; [19] rstudioapi_0.16.0   tools_4.5.0         ragg_1.3.3         </span></span>
<span><span class="co">#&gt; [22] bslib_0.8.0         evaluate_1.0.1      yaml_2.3.10        </span></span>
<span><span class="co">#&gt; [25] BiocManager_1.30.25 jsonlite_1.8.9      rlang_1.1.4        </span></span>
<span><span class="co">#&gt; [28] fs_1.6.4            htmlwidgets_1.6.4</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-liu_standr_2023" class="csl-entry">
Liu, Ning, Dharmesh D Bhuva, Ahmed Mohamed, Micah Bokelund, Arutha
Kulasinghe, Chin Wee Tan, and Melissa J Davis. 2023. <span>“<span class="nocase">standR</span>: Spatial Transcriptomic Analysis for
<span>GeoMx</span> <span>DSP</span> Data.”</span> <em>Nucleic Acids
Research</em>, November, gkad1026. <a href="https://doi.org/10.1093/nar/gkad1026" class="external-link">https://doi.org/10.1093/nar/gkad1026</a>.
</div>
<div id="ref-ritchie_limma_2015" class="csl-entry">
Ritchie, Matthew E., Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law,
Wei Shi, and Gordon K. Smyth. 2015. <span>“Limma Powers Differential
Expression Analyses for <span>RNA</span>-Sequencing and Microarray
Studies.”</span> <em>Nucleic Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Seung J. Kim, Marco Mura.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
